Sources/NIOSSL/SecurityFrameworkCertificateVerification.swift

```diff
@@ -47,11 +47,48 @@ extension SSLConnection {
             var trust: SecTrust? = nil
             var result: OSStatus
             let policy = SecPolicyCreateSSL(self.role! == .client, self.expectedHostname as CFString?)
-            result = SecTrustCreateWithCertificates(certificates as CFArray, policy, &trust)
+            result = SecTrustCreateWithCertificates(peerCertificates as CFArray, policy, &trust)
             guard result == errSecSuccess, let actualTrust = trust else {
                 throw NIOSSLError.unableToValidateCertificate
             }
 
+            // If there are additional trust roots then we need to add them to the SecTrust as anchors.
+            var additionalAnchorCertificates: [SecCertificate] = try self.parentContext.configuration.additionalTrustRoots.flatMap { trustRoots -> [NIOSSLCertificate] in
+                switch trustRoots {
+                case .file(let path):
+                    return try NIOSSLCertificate.fromPEMFile(path)
+                case .certificates(let certs):
+                    return certs
+                }
+            }.map {
+                guard let secCert = SecCertificateCreateWithData(nil, Data(try $0.toDERBytes()) as CFData) else {
+                    throw NIOSSLError.failedToLoadCertificate
+                }
+                return secCert
+            }
+            if !additionalAnchorCertificates.isEmpty {
+                // To use additional anchors _and_ the built-in ones we must reenable the built-in ones expicitly.
+                switch self.parentContext.configuration.trustRoots {
+                case .none, .some(.default):
+                    guard SecTrustSetAnchorCertificatesOnly(actualTrust, false) == errSecSuccess else {
+                        throw NIOSSLError.failedToLoadCertificate
+                    }
+                default:
+                    break
+                }
+                // We should add our additional anchors to any custom ones that already exist.
+                var existingCustomAnchorCertificates: CFArray? = nil
+                guard SecTrustCopyCustomAnchorCertificates(actualTrust, &existingCustomAnchorCertificates) == errSecSuccess else {
```

Comment:
This should not be necessary: `SecTrustSetAnchorCertificatesOnly(false)` should be sufficient to trust the built in anchors without us re-adding them.

Reply:
I had some reason for this, let me re-read the docs and see if my reasoning was flawed and comment back.

Reply:
OK, here is the documentation for `SecTrustSetAnchorCertificates` that led me to include this step...

> If you want to create a set of anchor certificates by modifying the default set, call the `SecTrustCopyAnchorCertificates(_:)` function to obtain the current set of anchor certificates, modify that set as you wish, and create a new array of certificates. Then call `SecTrustSetAnchorCertificates(_:_:)` with the modified array.
>
> Note that when you call the `SecTrustSetAnchorCertificates(_:_:)` function, you are effectively telling the evaluation function to use the anchor certificates in the specified array to evaluate trust regardless of any user-specified trust settings for those certificates. Furthermore, any intermediate certificates based on those anchor certificates are also trusted without consulting user trust settings.

The way I interpreted this was that, without this copy and mutation, we would potentially be removing some of the "built-in" certs from the set of anchors—Do I misunderstand this?

(I also realise that I am calling `SecTrustCopyCustomAnchorCertificates(_:)` (cf. `SecTrustCopyAnchorCertificates(_:)`) which may not be intentional.

I think I wanted to make sure that our use of additional anchors was truly additional w.r.t. what the platform was configured with at the time of the call. I'm not sure what "built-in" means in the docs so tried to make sure I propagated everything.

In light of the above docs, do you think I need to make some tweaks to this read-and-update above, or still think I should remove it altogether?

Reply:
> The way I interpreted this was that, without this copy and mutation, we would potentially be removing some of the "built-in" certs from the set of anchors—Do I misunderstand this?

Yes, you do, but it's an easy mistake to make. I believe the above advice is supposed to cover only _modifications_: that is, when you want to perform general CRUD operations to the anchors. We don't want to do that, we want only to supplement the default anchors. That's handled by `SecTrustSetAnchorCertificatesOnly`, whose documentation says:

> **anchorCertificatesOnly**
> If true, disables trusting any anchors other than the ones passed in with the SecTrustSetAnchorCertificates(_:_:) function.  If false, the built-in anchor certificates are also trusted. If SecTrustSetAnchorCertificates(_:_:) is called and SecTrustSetAnchorCertificatesOnly(_:_:) is not called, only the anchors explicitly passed in are trusted.

Setting this to `false` in combination with a call to `SecTrustSetAnchorCertificates` should achieve our intended effect.

Reply:
OK I understand that `SecTrustSetAnchorCertificatesOnly(false)` and `SecTrustSetAnchorCertificates(...)` will mean that we get the "built-in anchor certs" and the ones we pass in in the `SecTrustSetAnchorCertificates(...)` call.

But what about the "custom anchors"? I was worried that these might inadvertently be discarded by these two calls from the verification, which I now remember is why I used `SecTrustCopy**Custom**AnchorCertificates`.

Shall we assume that we do not want these (if any have been configured)?

The docs for `SecTrustCopyCustomAnchorCertificates` state:

> You can use the `SecTrustSetAnchorCertificates(_:_:)` function to set custom anchor certificates.

which led me to think that certs passed into `SecTrustSetAnchorCertificates(_:_:)` are not considered "built-in" w.r.t. the  `SecTrustSetAnchorCertificatesOnly(false)` call?

Reply:
Certs passed into `SecTrustSetAnchorCertificates` are not built-in, they _are_ the custom anchors. So we don't need to preserve any: there is nowhere else they can have come from but us.

Reply:
Right. I think you have unlocked something in my understanding here... These custom certs are only custom to the SecTrust that we have created ourselves? They are not representing things that a user may have gone and done in Keychain?

If the above statement of my understanding is now correct then I realise now that I need to be removing this copy.

Reply:
Addressed in https://github.com/apple/swift-nio-ssl/pull/284/commits/ed01d7ad612f3401265d99672e9aa2efcc522f79.

Will leave this comment for you to resolve, @Lukasa, so you can validate my understanding in my previous comment.

Reply:
Yup, that understanding is correct.